platforms/core-configuration/graph-serialization/src/main/kotlin/org/gradle/internal/serialize/graph/Codec.kt

```diff
@@ -97,6 +98,19 @@ interface ReadContext : IsolateContext, MutableIsolateContext, Decoder {
 }
 
 
+interface MutableReadContext : ReadContext {
+    /**
+     * Sets a client specific property value that can be queried via [getSingletonProperty].
+     */
+    fun <T : Any> setSingletonProperty(singletonProperty: T)
+}
+
+
+interface CloseableReadContext : MutableReadContext, AutoCloseable {
+    fun finish()
```

Comment:
💭  We should explain why we need `finish` when the interface already extends `AutoCloseable`.

There seems to be a single place we call this function, which also happens to be inside the `useToRun {}` call, which will close the context at the end of the operation. There is also a call to the `runReadOperation` which is supposed to exhaust all pending operation, IIUC.

Maybe we can get rid of `finish` altogether?